### PR TITLE
5 Layer Prediction Structure

### PR DESCRIPTION
**Pull request by anaghdin**
Add the support for 5-Layer (mini Gop of 16 frames). By default 5L is used
run to run fix

### CODE REVIEW

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 1]
```diff
@@ -300,7 +300,296 @@ EbErrorType ReleasePrevPictureFromReorderQueue(
...
+***************************************************************************************************/
+EbErrorType update_base_layer_reference_queue_dependent_count(
+    PictureDecisionContext_t        *context_ptr,
+    EncodeContext_t                 *encode_context_ptr,
+    SequenceControlSet_t            *sequence_control_set_ptr,
+    uint32_t                           miniGopIndex) {
+
+    EbErrorType return_error = EB_ErrorNone;
+
+    PaReferenceQueueEntry_t         *inputEntryPtr;
```
Comment (hassount) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: snake_case here and the function

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 2]
```diff
@@ -300,7 +300,296 @@ EbErrorType ReleasePrevPictureFromReorderQueue(
...
+                inputEntryPtr->depList1Count = inputEntryPtr->list1.listCount;
+                inputEntryPtr->dependentCount = inputEntryPtr->depList0Count + inputEntryPtr->depList1Count - dependantListRemovedEntries;
+
+            }
+            else {
+
+                // Modify Dependent List0
+                depListCount = inputEntryPtr->list0.listCount;
+                for (depIdx = 0; depIdx < depListCount; ++depIdx) {
+
```
Comment (hassount) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: too many spaces between lines

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 3]
```diff
@@ -710,15 +999,233 @@ void  Av1GenerateRpsInfo(
 
 
     }
-    else
+#if NEW_PRED_STRUCT
```
Comment (hassount) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: too many spaces

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 4]
```diff
@@ -710,15 +999,233 @@ void  Av1GenerateRpsInfo(
...
+    //         1     3    5      7    9     11     13      15
+    //            2          6           10            14
+    //                 4                        12
+    //                              8
+    //base0:0                                               base1:16
+    const uint8_t  base0Idx = context_ptr->miniGopToggle ? 0 : 3; //Base layer for prediction from past
+    const uint8_t  base1Idx = context_ptr->miniGopToggle ? 3 : 0; //Base layer for prediction from future
+    const uint8_t  layer1Idx = 1;
+    const uint8_t  layer2Idx = 2;
+    const uint8_t  layer3Idx1 = 4;
```
Comment (joelsole) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: snake case

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 5]
```diff
@@ -300,7 +300,296 @@ EbErrorType ReleasePrevPictureFromReorderQueue(
...
+    // Loop over all mini GOPs
+    mini_gop_index = 0;
+    while (mini_gop_index < MINI_GOP_MAX_COUNT) {
+
+        // Only for a valid mini GOP
+        if (GetMiniGopStats(mini_gop_index)->endIndex < encode_context_ptr->pre_assignment_buffer_count && context_ptr->miniGopActivityArray[mini_gop_index] == EB_FALSE) {
+
+            context_ptr->miniGopStartIndex[context_ptr->totalNumberOfMiniGops] = GetMiniGopStats(mini_gop_index)->startIndex;
+            context_ptr->miniGopEndIndex[context_ptr->totalNumberOfMiniGops] = GetMiniGopStats(mini_gop_index)->endIndex;
+            context_ptr->miniGopLenght[context_ptr->totalNumberOfMiniGops] = GetMiniGopStats(mini_gop_index)->lenght;
```
Comment (joelsole) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]: on a clean-up, let's replace these lenght by length

File: Source/Lib/Codec/EbPictureDecisionProcess.c [hunk 6]
```diff
@@ -300,7 +300,296 @@ EbErrorType ReleasePrevPictureFromReorderQueue(
...
+/***************************************************************************************************
+* If a switch happens, then update the RPS of the base layer frame separating the 2 different prediction structures
+* Clean up the reference queue dependant counts of the base layer frame separating the 2 different prediction structures
+*
+***************************************************************************************************/
+EbErrorType update_base_layer_reference_queue_dependent_count(
+    PictureDecisionContext_t        *context_ptr,
+    EncodeContext_t                 *encode_context_ptr,
+    SequenceControlSet_t            *sequence_control_set_ptr,
+    uint32_t                           miniGopIndex) {
```
Comment (kylophone) [inline Source/Lib/Codec/EbPictureDecisionProcess.c]:
```
if (!context_ptr || !encode_context_ptr || ! sequence_control_set_ptr)
    return EB_ErrorBadParameter
```

**Comment by joelsole** [comment]
lgtm

**Pull request merged**